### ISSUE TITLE
i am still wondering why prediction is just white

### ISSUE THREAD

**Issue by bamtu** [tsugumi-sys/SA-ConvLSTM-Pytorch#43]
i trained this model last night with Adam lr=0.0001, with 
model = SAMSeq2Seq(attention_hidden_dims=1, num_channels=1, num_kernels=1, input_seq_length=10,
kernel_size=3, padding='same', activation="relu", return_sequences = True, out_channels = 1,
frame_size=(64, 64), num_layers=1, weights_initializer='he').to()

and after 100 epoch, the prediction is below. 

<img width="970" alt="image" src="https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/34215754/85c0e0b0-0ea0-44e4-ae67-f4a078ef9610">

can you explain this?

i am very thankful for your swiftness and attention..

**Comment by tsugumi-sys** [comment]
@bamtu 
Hmmm :(
Can you check the learning curves?
(learning data is saved into "./tmp/metrics.csv" if you're using default arguments.)

**Comment by bamtu** [comment]
and.... can you check the result after one epoch at num_kernels=40 ?

![test-case49](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/34215754/4fa8148b-fc69-4276-b855-6970bb0b1c4c)

**Comment by bamtu** [comment]
i tried to train convLSTM after one epoch.
![test-case0](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/34215754/550ce121-19bf-43cc-abce-297c01ffd14b)
every prediction is same and only white. 

In conclusion, SA-convLSTM and standard ConvLSTM prediction are weird.

**Comment by bamtu** [comment]
![image](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/34215754/776c0ae7-30f6-409f-b97f-3f68c92aebd6)
and prediction after one epoch of my own code(ConvLSTM)(from another github repository)

**Comment by tsugumi-sys** [comment]
@bamtu 
I see, my model is weird.
Can you share the ConvLSTM respository?
And I also want to know your optimizer and loss function settings :)

**Comment by tsugumi-sys** [comment]
And you can try increasing `num_layers` and `attention_hidden_dims` parameters for `SAMSeq2Seq` (e.g. num_layers=4, attention_hidden_dims=4).
In my another research (rainfall prediction task), SAMSeq2Seq achieved higher accuracy.

I'll check my code and train again for MovingMNIST dataset.

**Comment by bamtu** [comment]
here is a jupyter notebook i tried. i don't know where it comes and i edited a little
[EX_ConvLSTM_Code_1layer.zip](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/files/13825263/EX_ConvLSTM_Code_1layer.zip)

**Comment by bamtu** [comment]
and here is another SA-ConvLSTM implementation but can't apply movingMNIST dataset.

https://github.com/jerrywn121/TianChi_AIEarth

**Comment by bamtu** [comment]
more implementation : https://github.com/MinNamgung/sa_convlstm
https://github.com/hyona-yu/SA-convlstm

**Comment by tsugumi-sys** [comment]
@bamtu 
I'm going to check your codes and my model implementation today :)
Please wait a moment :)

**Comment by tsugumi-sys** [comment]
@bamtu 
I'm sorry but I'm busy today :( I'll check tomorrow

**Comment by tsugumi-sys** [comment]
@bamtu 
I found that the forecasting process of your model (https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/files/13825263/EX_ConvLSTM_Code_1layer.zip) is different from that of my model.

Your model returns only one frame, so the input data is updated sequentially to predict next 12 frames.

```python
# Seq2Seq implementation in notebook
    def forward(self, X):

        # Forward propagation through all the layers
        output = self.sequential(X)

        # Return only the last output frame
        output = self.conv(output[:,:,-1])

        return nn.Sigmoid()(output)
```
```python
MovingMNIST = np.load('/content/drive/MyDrive/Colab Notebooks/mnist_test_seq.npy').transpose(1, 0, 2, 3)

# Train, Test, Validation splits
test_data = MovingMNIST[9000:9003]

# Test Data Loader
test_loader = DataLoader(test_data, batch_size=3, collate_fn=collate_test)

# Get a batch
batch, target = next(iter(test_loader))

# Initialize output sequence
output = np.zeros(target.shape, dtype=np.uint8)

# print(model)
# Loop over timesteps
for timestep in range(target.shape[1]):
   # update inputs sequentially here
    input = batch[:,:,timestep:timestep+10]
    output[:,timestep] = model(input).squeeze(1).cpu().detach().numpy() * 255.0
```

**Comment by tsugumi-sys** [comment]
In other words, your model predicts next one frame from inputs, but my model predicts next 10 frames from inputs.

So your model's prediction task is simpler than that of my model, thus we cannot compare the prediction results of both models.

**Comment by tsugumi-sys** [comment]
The forward implementation of my model is different :)
https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/blob/d97be78fc0155c3e17ad51e3d81750d9bb42c968/convlstm/seq2seq.py#L108-L109

**Comment by bamtu** [comment]
i get it. so... you want me to predict one by one following your code? 

Even if you're right, the result is still weird. 

![test-case0](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/34215754/a0d9a1fc-3ba3-476a-8106-e1ee2cd0be52)
![test-case1](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/34215754/22e41387-0224-48f3-a71c-f9f2c64a8e45)
![test-case2](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/34215754/cce31663-e63a-4098-af7c-b91e825dcdef)

**Comment by bamtu** [comment]
For the proper results, What should i do?

**Comment by tsugumi-sys** [comment]
> i get it. so... you want me to predict one by one following your code?

I think we need to check other setting like loss function, optimizer, e.t.c.

> For the proper results, What should i do?

It depends on your inference task.
Do you want the model to predict like the notebook code (just predict next one frame), or predict more than one frames?

**Comment by bamtu** [comment]
i want the model to predict like my notebook code if possible. 

but, Isn't it an issue that output is only white? i think someone should correct it..

**Comment by tsugumi-sys** [comment]
@bamtu 
I see. Anyway, I need to check the white output so please wait a moment please :)
(I think it'll take few days or weeks.)

**Comment by tsugumi-sys** [comment]
@bamtu 
I've fixed the white output issue.
I added an example of the ConvLSTM Seq2Seq model with the same model parameters and loss.

![test-case0](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/61897166/37315812-3beb-4bb3-9933-e3c42b28bb19)


Please pull the latest code and try the following command :)

```bash
python -m examples.moving_mnist_convlstm
```

**Comment by tsugumi-sys** [comment]
I'm going to SAConvLSTM next :)

**Comment by tsugumi-sys** [comment]
It's done :)
https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/blob/main/examples/moving_mnist_self_attention_memory_convlstm.py
